internal/cmd/gocdk/internal/docker/docker.go

```diff
@@ -0,0 +1,227 @@
+	// See https://godoc.org/github.com/docker/distribution/reference
+	Tag string
+	// Digest is the content-based hash of the image.
+	// It may be empty.
+	Digest string
+	// CreatedAt is the time the image was built.
+	CreatedAt time.Time
+}
+
+// ListImages lists images the Docker daemon has stored locally that match filterRef.
+func (c *Client) ListImages(ctx context.Context, filterRef string) ([]*Image, error) {
+	args := []string{"images", "--digests", "--format={{json .}}"}
+	if filterRef != "" {
+		args = append(args, filterRef)
+	}
+	cmd := exec.CommandContext(ctx, "docker", args...)
+	cmd.Env = c.env
+	out, err := run(cmd)
+	if err != nil {
+		return nil, xerrors.Errorf("list docker images: %w", err)
+	}
+	lf := []byte("\n")
+	out = bytes.TrimSuffix(out, lf)
+	var images []*Image
+	for _, line := range bytes.Split(out, lf) {
+		var img struct {
+			ID         string
+			Repository string
+			Tag        string
+			Digest     string
+			CreatedAt  string
+		}
+		if err := json.Unmarshal(line, &img); err != nil {
+			return nil, xerrors.Errorf("list docker images: parse output: %w", err)
+		}
+		createdAt, _ := time.Parse("2006-01-02 15:04:05 -0700 MST", img.CreatedAt)
+		images = append(images, &Image{
+			ID:         cleanNone(img.ID),
+			Repository: cleanNone(img.Repository),
+			Tag:        cleanNone(img.Tag),
+			Digest:     cleanNone(img.Digest),
+			CreatedAt:  createdAt,
+		})
+	}
+	return images, nil
+}
+
+func cleanNone(s string) string {
+	if s == "<none>" {
+		return ""
+	}
+	return s
+}
+
+// RunOptions holds optional parameters for starting a Docker container.
+type RunOptions struct {
+	Args         []string
+	Env          []string
+	RemoveOnExit bool
+	Publish      []string
+}
+
+// Start starts a Docker container and returns its ID.
+func (c *Client) Start(ctx context.Context, imageRef string, opts *RunOptions) (string, error) {
+	args := []string{"run", "--detach"}
+	if opts != nil {
+		if opts.RemoveOnExit {
+			args = append(args, "--rm")
+		}
+		for _, e := range opts.Env {
+			args = append(args, "--env", e)
+		}
+		for _, p := range opts.Publish {
+			args = append(args, "--publish", p)
+		}
+	}
+	args = append(args, "--", imageRef)
+	if opts != nil {
+		args = append(args, opts.Args...)
+	}
+
+	cmd := exec.CommandContext(ctx, "docker", args...)
+	cmd.Env = c.env
+	out, err := run(cmd)
+	if err != nil {
+		return "", xerrors.Errorf("docker run %s: %w", imageRef, err)
+	}
+	out = bytes.TrimSuffix(out, []byte("\n"))
+	if len(out) == 0 {
+		return "", xerrors.Errorf("docker run %s: docker did not return a container ID", imageRef)
+	}
+	return string(out), nil
+}
+
+// Build builds a Docker image from a source directory, optionally tagging it
+// with the given image reference. Output is streamed to the buildOutput writer.
+func (c *Client) Build(ctx context.Context, imageRef string, dir string, buildOutput io.Writer) error {
+	cmd := exec.CommandContext(ctx, "docker", "build", "-t="+imageRef, "--", dir)
```

Comment:
What is the `--` (5th argument) doing here?  If I understand correctly, this constructs something that would look on a command like like `docker build -t=myimageref:tag -- path/somewhere, but I don't see an example of that usage on https://docs.docker.com/engine/reference/commandline/build/,  so I'm a little confused.

Reply:
It prevents the subsequent arguments being interpreted as command-line flags, so if for instance the user tries to build a directory that starts with a hyphen, the `docker build` command would understand it properly.

I don't have a Docker-specific reference that documents this behavior, but is a behavior emulated by [GNU getopt](https://www.gnu.org/software/libc/manual/html_node/Using-Getopt.html)-compatible flag-parsing libraries.